File: Neetcode/PaliPartition.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        out = []
        res = []
        def pali(st):
            i = 0
            j = len(st)-1
            while i <= j:
                if st[i] != st[j]:
                    return False
                i += 1
                j -= 1
            return True
        def bt(i):
            if i >= len(s):
                out.append(res.copy())
                return
            for j in range(i, len(s)):
                if pali(s[i:j+1]):
                    res.append(s[i:j+1])
                    bt(j+1)
                    res.pop()
        bt(0)
        return out

File: Neetcode/PaliPartition.cpp
class Solution {
public:
    vector<vector<string>> out;
    vector<string> res;
    bool pali(string s, int i, int j){
        while (i < j){
            if (s[i] != s[j]){
                return false;
            }
            i += 1;
            j -= 1;
        }
        return true;
    }
    void bt(string s, int i){
        if (i >= s.size()){
            out.push_back(res);
            return;
        }
        for(int j=i; j<s.size(); j++){
            if (pali(s, i, j)){
                res.push_back(s.substr(i, j-i+1));
                bt(s, j+1);
                res.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        bt(s, 0);
        return out;
    }
};
